frontend/src/metabase/admin/databases/containers/DatabaseEditApp.unit.spec.js

```diff
@@ -53,9 +59,43 @@ async function setup({ cachingEnabled = false } = {}) {
   });
 
   await waitForElementToBeRemoved(() => screen.queryByText("Loading..."));
+
+  return { mockEventListener };
 }
 
 describe("DatabaseEditApp", () => {
+  describe("Database connections", () => {
+    afterEach(() => {
+      jest.restoreAllMocks();
+    });
+
+    it("should trigger beforeunload event when database connection is edited", async () => {
+      const { mockEventListener } = await setup();
+
+      const databaseForm = await screen.findByTestId("database-name-field");
+
+      // Workaround for a known bug in Formik when testing with react-testing-library
+      // https://github.com/jaredpalmer/formik/issues/1543#issuecomment-547501926
+      // eslint-disable-next-line testing-library/no-unnecessary-act
+      await act(async () => {
+        await userEvent.type(databaseForm, "Test database");
+      });
+
+      const mockEvent = callMockEvent(mockEventListener, "beforeunload");
+
+      expect(mockEvent.preventDefault).toHaveBeenCalled();
+      expect(mockEvent.returnValue).toBe(BEFORE_UNLOAD_UNSAVED_MESSAGE);
+    });
```

Comment:
`waitFor` should let us avoid using `act` here. What if we do something like that:

```ts
import { waitFor } from "__support__/ui";

// ...

const { mockEventListener } = await setup();
const databaseForm = await screen.findByTestId("database-name-field");

userEvent.type(databaseForm, "Test database");

await waitFor(() => {
  const mockEvent = callMockEvent(mockEventListener, "beforeunload");
  expect(mockEvent.preventDefault).toHaveBeenCalled();
});
expect(mockEvent.returnValue).toBe(BEFORE_UNLOAD_UNSAVED_MESSAGE);
```